homeassistant/components/fan/reproduce_state.py

```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
+                    data[attr] = state.attributes[attr]
+            service_calls[SERVICE_TURN_ON] = data
+        else:
+            # If the fan is already on, we need to set speed or mode
+            # based on the state.
+            #
+            # Speed and preset mode are mutually exclusive, so one of
+            # them is always going to be stored as None. If we were to
+            # try to set it, it will raise an error. So instead we
+            # only set the one that is non-None.
+            for attr, service in SPEED_AND_MODES_ATTRIBUTES.items():
+                if (
+                    attr in state.attributes
+                    and state.attributes[attr] is not None
+                    and not check_attr_equal(
```

Comment:
Similar to the suggestion for the `None` check above, lets pull out a variable for the attributes that differences, e.g., something like this:
```
updated_attributes = {k: v for k,v in state_attributes.items() if state_attributes.get(k) != v }
```

Then that can be used to simplify building the updates.

Reply:
(I'm not sure we even need `check_attr_equal` given its a pretty trivial check)